File: Sources/GatewayCore/AudioIO.swift
import Foundation
import AVFoundation

/// Wav in and out, via AVFoundation -- no extra dependencies. Everything
/// internal to the pipeline is mono float32 at **24 kHz**, which survived the
/// engine change: Qwen3-TTS-12Hz outputs 24 kHz too (the 12 Hz names the codec
/// frame rate, not the audio), so nothing here resamples.
///
/// Lives in GatewayCore rather than GatewayTTS because it has nothing to do
/// with the synthesiser: it is plain file I/O, and `gfcheck` needs it to read
/// rendered audio back and hold it to account. That separation matters more
/// now, not less -- once mlx-swift lands, anything importing GatewayTTS can
/// only be built by `xcodebuild`, and dragging the check binary across that
/// line to reach a wav reader would be the wrong trade twice over.
public enum AudioIO {
    public static let sampleRate = 24000.0

    public struct FileMetadata: Equatable, Sendable {
        public var seconds: Double
        public var sampleRate: Double
        public var channels: Int

        public init(seconds: Double, sampleRate: Double, channels: Int) {
            self.seconds = seconds; self.sampleRate = sampleRate; self.channels = channels
        }
    }

    /// Reads the container rather than trusting a catalog entry or filename.
    public static func metadata(_ url: URL) throws -> FileMetadata {
        let file = try AVAudioFile(forReading: url)
        let format = file.fileFormat
        return FileMetadata(seconds: Double(file.length) / format.sampleRate,
                            sampleRate: format.sampleRate,
                            channels: Int(format.channelCount))
    }

    public static func loadStereo(_ url: URL) throws -> StereoAudio {
        let file = try AVAudioFile(forReading: url)
        let format = file.processingFormat
        let frames = AVAudioFrameCount(file.length)
        guard let buffer = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: frames) else {
            throw err("stereo buffer alloc failed")
        }
        try file.read(into: buffer)
        guard let channels = buffer.floatChannelData, buffer.frameLength > 0 else {
            throw err("audio is not readable as float PCM")
        }
        let count = Int(buffer.frameLength)
        let left = Array(UnsafeBufferPointer(start: channels[0], count: count))
        let right = format.channelCount > 1
            ? Array(UnsafeBufferPointer(start: channels[1], count: count)) : left
        return StereoAudio(sampleRate: format.sampleRate, left: left, right: right)
    }

    /// Any readable audio file -> mono float32 at 24 kHz.
    public static func loadMono24k(_ url: URL) throws -> [Float] {
        let file = try AVAudioFile(forReading: url)
        let inFmt = file.processingFormat
        let frames = AVAudioFrameCount(file.length)
        guard let inBuf = AVAudioPCMBuffer(pcmFormat: inFmt, frameCapacity: frames)
        else { throw err("buffer alloc failed") }
        try file.read(into: inBuf)

        guard let outFmt = AVAudioFormat(commonFormat: .pcmFormatFloat32,
                                         sampleRate: sampleRate, channels: 1,
                                         interleaved: false),
              let conv = AVAudioConverter(from: inFmt, to: outFmt)
        else { throw err("converter setup failed") }
        let outFrames = AVAudioFrameCount(Double(frames) * sampleRate / inFmt.sampleRate + 1024)
        guard let outBuf = AVAudioPCMBuffer(pcmFormat: outFmt, frameCapacity: outFrames)
        else { throw err("out buffer alloc failed") }
        var fed = false
        var convErr: NSError?
        conv.convert(to: outBuf, error: &convErr) { _, status in
            if fed { status.pointee = .endOfStream; return nil }
            fed = true; status.pointee = .haveData; return inBuf
        }
        if let convErr { throw convErr }
        guard let ch = outBuf.floatChannelData else { throw err("no channel data") }
        return Array(UnsafeBufferPointer(start: ch[0], count: Int(outBuf.frameLength)))
    }

    /// Mono float32 -> 16-bit PCM wav.
    public static func writeWav(_ samples: [Float], to url: URL,
                                sampleRate: Int = 24000) throws {
        var data = Data()
        func put(_ s: String) { data.append(s.data(using: .ascii)!) }
        func put32(_ v: UInt32) { withUnsafeBytes(of: v.littleEndian) { data.append(contentsOf: $0) } }
        func put16(_ v: UInt16) { withUnsafeBytes(of: v.littleEndian) { data.append(contentsOf: $0) } }
        let payload = samples.count * 2
        put("RIFF"); put32(UInt32(36 + payload)); put("WAVE")
        put("fmt "); put32(16); put16(1); put16(1)
        put32(UInt32(sampleRate)); put32(UInt32(sampleRate * 2)); put16(2); put16(16)
        put("data"); put32(UInt32(payload))
        var pcm = [Int16](repeating: 0, count: samples.count)
        for (i, s) in samples.enumerated() {
            pcm[i] = Int16(max(-1, min(1, s)) * 32767)
        }
        pcm.withUnsafeBytes { data.append(contentsOf: $0) }
        try data.write(to: url, options: .atomic)
    }

    /// The same, in two channels.
    ///
    /// The bed is stereo by design -- the binaural pair is the whole point, and
    /// surf and tuning are widened deliberately -- so auditioning it through the
    /// mono writer would throw away the thing being auditioned.
    public static func writeWavStereo(left: [Float], right: [Float], to url: URL,
                                      sampleRate: Int = 24000) throws {
        let n = min(left.count, right.count)
        var data = Data()
        func put(_ s: String) { data.append(s.data(using: .ascii)!) }
        func put32(_ v: UInt32) { withUnsafeBytes(of: v.littleEndian) { data.append(contentsOf: $0) } }
        func put16(_ v: UInt16) { withUnsafeBytes(of: v.littleEndian) { data.append(contentsOf: $0) } }
        let payload = n * 2 * 2                      // frames x channels x 16-bit
        put("RIFF"); put32(UInt32(36 + payload)); put("WAVE")
        put("fmt "); put32(16); put16(1); put16(2)
        put32(UInt32(sampleRate)); put32(UInt32(sampleRate * 4)); put16(4); put16(16)
        put("data"); put32(UInt32(payload))
        var pcm = [Int16](repeating: 0, count: n * 2)
        for i in 0..<n {
            pcm[i * 2]     = Int16(max(-1, min(1, left[i])) * 32767)
            pcm[i * 2 + 1] = Int16(max(-1, min(1, right[i])) * 32767)
        }
        pcm.withUnsafeBytes { data.append(contentsOf: $0) }
        try data.write(to: url, options: .atomic)
    }

    private static func err(_ m: String) -> NSError {
        NSError(domain: "AudioIO", code: 1, userInfo: [NSLocalizedDescriptionKey: m])
    }
}

/// Measuring rendered audio, so a broken render can be *found* rather than
/// listened for.
///
/// A whole library was rendered with speech followed by a long silence tail --
/// files that look finished on disk, play back as "the voice stops mid-sentence
/// and picks up two lines later", and nothing in the pipeline noticed. The
/// signature is silence that no `pause` or `hold` in the script asked for.
public enum AudioProbe {
    public struct RenderQuality: Sendable {
        public let seconds: Double
        public let peak: Float
        public let clippedSamples: Int
        public let nonFiniteSamples: Int
        public let leadingQuietSeconds: Double
        public let trailingQuietSeconds: Double

        /// This is intentionally a small, mechanical contract. It does not
        /// claim that speech sounds good; it rejects the file-level defects we
        /// can prove without pretending an acoustic metric has ears.
        public var safe: Bool {
            seconds > 0 && clippedSamples == 0 && nonFiniteSamples == 0
                && leadingQuietSeconds >= RenderPlan.speechEdgeQuietSeconds * 0.95
                && trailingQuietSeconds >= RenderPlan.speechEdgeQuietSeconds * 0.95
        }
    }

    /// Peak/clipping and the two file edges of a rendered speech unit.
    /// `preparedSpeechPart` guarantees the edge quiet; this reads it back so a
    /// future refactor cannot silently remove the guarantee.
    public static func renderQuality(_ samples: [Float],
                                     sampleRate: Double = AudioIO.sampleRate,
                                     quietThreshold: Float = RenderPlan.speechEdgeThreshold)
        -> RenderQuality {
        var peak: Float = 0
        var clipped = 0
        var nonFinite = 0
        for sample in samples {
            guard sample.isFinite else { nonFinite += 1; continue }
            let magnitude = abs(sample)
            peak = max(peak, magnitude)
            if magnitude >= 0.999 { clipped += 1 }
        }
        var leading = 0
        while leading < samples.count,
              samples[leading].isFinite,
              abs(samples[leading]) < quietThreshold { leading += 1 }
        var trailing = 0
        while trailing < samples.count,
              samples[samples.count - 1 - trailing].isFinite,
              abs(samples[samples.count - 1 - trailing]) < quietThreshold { trailing += 1 }
        return RenderQuality(
            seconds: Double(samples.count) / sampleRate,
            peak: peak,
            clippedSamples: clipped,
            nonFiniteSamples: nonFinite,
            leadingQuietSeconds: Double(leading) / sampleRate,
            trailingQuietSeconds: Double(trailing) / sampleRate)
    }

    /// Runs of near-silence, as (start, duration) in seconds.
    public static func silences(_ samples: [Float], sampleRate: Double = AudioIO.sampleRate,
                                threshold: Float = 0.005, window: Double = 0.05) -> [(Double, Double)] {
        let w = max(1, Int(window * sampleRate))
        let frames = stride(from: 0, to: samples.count, by: w).map { i -> Bool in
            var peak: Float = 0
            for s in samples[i..<min(i + w, samples.count)] { peak = max(peak, abs(s)) }
            return peak < threshold
        }
        var out: [(Double, Double)] = []
        var i = 0
        while i < frames.count {
            guard frames[i] else { i += 1; continue }
            var j = i
            while j < frames.count, frames[j] { j += 1 }
            out.append((Double(i) * window, Double(j - i) * window))
            i = j
        }
        return out
    }

    /// Silence the script never asked for: a run longer than `minRun` that does
    /// not match any `pause` or `hold` in the body it was rendered from.
    ///
    /// Tolerant on purpose -- a rendered line ends with its own natural breath,
    /// so a scripted 10 s pause reads as a little more than 10 s. What it is
    /// looking for is tens of seconds of nothing where words should be.
    public static func unexplainedSilence(samples: [Float], doc: ScriptDoc,
                                          sampleRate: Double = AudioIO.sampleRate,
                                          minRun: Double = 8, tolerance: Double = 3)
        -> [(start: Double, seconds: Double)] {
        // Consecutive written silences are *heard* as one, so they must be
        // matched as one. `pause 8` followed by `hold 60` is 68 seconds of
        // unbroken quiet in the render, and comparing that against 8 and 60
        // separately reported a 68-second hole nobody wrote -- a false alarm on
        // the one check that exists to catch real ones, which is the worst
        // place to have a false alarm.
        var scripted: [Double] = []
        var run = 0.0
        for step in doc.steps {
            switch step.kind {
            case .pause, .hold, .media:
                run += step.seconds
            case .say:
                if run > 0 { scripted.append(run); run = 0 }
            default: break
            }
        }
        if run > 0 { scripted.append(run) }

        return silences(samples, sampleRate: sampleRate)
            .filter { silence in
                silence.1 >= minRun && !scripted.contains { abs(silence.1 - $0) <= tolerance }
            }
            .map { (start: $0.0, seconds: $0.1) }
    }
}

File: cross-platform/src/core/audioIO.ts
/**
 * Wav in and out, ported from `Sources/GatewayCore/AudioIO.swift`.
 *
 * Everything internal to the pipeline is mono float32 at **24 kHz**, and the
 * files on disk are 16-bit PCM. This is plain file I/O and belongs beside the
 * rest of the ported core rather than with the synthesiser, for the same
 * reason it does on the Swift side: a check needs to read rendered audio back
 * and hold it to account without dragging the speech engine in with it.
 *
 * **The two conversions are deliberately asymmetric, because Swift's are.**
 * Writing multiplies by 32767 and clamps; reading divides by 32768, which is
 * what `AVAudioFile` does handing a 16-bit file to a float buffer. Using one
 * constant for both would be tidier and would put every sample about 3e-5 away
 * from what the macOS build reads — measured, when a first draft did exactly
 * that.
 */
import { readFileSync, writeFileSync } from "fs";
import { makeResampler } from "../main/resample.js";

export const sampleRate = 24000;

export interface StereoAudio { sampleRate: number; left: Float32Array; right: Float32Array }
export interface FileMetadata { seconds: number; sampleRate: number; channels: number }

interface RawWav { sampleRate: number; channels: number; bits: number; frames: number; data: Buffer }

/** Walks the RIFF chunks rather than assuming `fmt ` and `data` sit at fixed
 *  offsets — they usually do, and a file carrying a `LIST` chunk would read as
 *  noise if that were assumed. */
function readRiff(path: string): RawWav {
  const b = readFileSync(path);
  if (b.length < 12 || b.toString("ascii", 0, 4) !== "RIFF" || b.toString("ascii", 8, 12) !== "WAVE") {
    throw new Error(`${path} is not a RIFF/WAVE file`);
  }
  let offset = 12, channels = 0, rate = 0, bits = 0, format = 0;
  let data: Buffer | undefined;
  while (offset + 8 <= b.length) {
    const id = b.toString("ascii", offset, offset + 4);
    const size = b.readUInt32LE(offset + 4);
    const body = offset + 8;
    if (id === "fmt ") {
      format = b.readUInt16LE(body);
      channels = b.readUInt16LE(body + 2);
      rate = b.readUInt32LE(body + 4);
      bits = b.readUInt16LE(body + 14);
    } else if (id === "data") {
      data = b.subarray(body, Math.min(b.length, body + size));
    }
    offset = body + size + (size % 2);          // chunks are word-aligned
  }
  if (data === undefined) throw new Error(`${path} has no data chunk`);
  if (format !== 1 || bits !== 16) {
    throw new Error(`${path} is not 16-bit PCM (format ${format}, ${bits}-bit)`);
  }
  if (channels < 1) throw new Error(`${path} declares no channels`);
  return { sampleRate: rate, channels, bits, frames: Math.floor(data.length / 2 / channels), data };
}

/** Reads the container rather than trusting a catalogue entry or a filename. */
export function metadata(path: string): FileMetadata {
  const w = readRiff(path);
  return { seconds: w.frames / w.sampleRate, sampleRate: w.sampleRate, channels: w.channels };
}

const toFloat = (raw: Buffer, index: number): number => raw.readInt16LE(index * 2) / 32768;

export function loadStereo(path: string): StereoAudio {
  const w = readRiff(path);
  const left = new Float32Array(w.frames);
  const right = new Float32Array(w.frames);
  for (let i = 0; i < w.frames; i++) {
    left[i] = toFloat(w.data, i * w.channels);
    // A mono file read as stereo is the same signal in both ears, not silence
    // on the right — matching `loadStereo`'s own fallback.
    right[i] = w.channels > 1 ? toFloat(w.data, i * w.channels + 1) : left[i]!;
  }
  return { sampleRate: w.sampleRate, left, right };
}

/** Any readable wav to mono float32 at 24 kHz. */
export function loadMono24k(path: string): Float32Array {
  const w = readRiff(path);
  const mono = new Float32Array(w.frames);
  for (let i = 0; i < w.frames; i++) {
    let sum = 0;
    for (let c = 0; c < w.channels; c++) sum += toFloat(w.data, i * w.channels + c);
    mono[i] = sum / w.channels;
  }
  if (w.sampleRate === sampleRate) return mono;
  return makeResampler(w.sampleRate, sampleRate).run(mono);
}

function header(channels: number, frames: number, rate: number): Buffer {
  const payload = frames * channels * 2;
  const b = Buffer.alloc(44);
  b.write("RIFF", 0, "ascii");
  b.writeUInt32LE(36 + payload, 4);
  b.write("WAVE", 8, "ascii");
  b.write("fmt ", 12, "ascii");
  b.writeUInt32LE(16, 16);
  b.writeUInt16LE(1, 20);                        // PCM
  b.writeUInt16LE(channels, 22);
  b.writeUInt32LE(rate, 24);
  b.writeUInt32LE(rate * 2 * channels, 28);      // byte rate
  b.writeUInt16LE(2 * channels, 32);             // block align
  b.writeUInt16LE(16, 34);
  b.write("data", 36, "ascii");
  b.writeUInt32LE(payload, 40);
  return b;
}

const toPCM = (v: number): number => Math.trunc(Math.max(-1, Math.min(1, v)) * 32767);

export function writeWav(samples: Float32Array, path: string, rate = sampleRate): void {
  const body = Buffer.alloc(samples.length * 2);
  for (let i = 0; i < samples.length; i++) body.writeInt16LE(toPCM(samples[i]!), i * 2);
  writeFileSync(path, Buffer.concat([header(1, samples.length, rate), body]));
}

/** The same, in two channels. The bed is stereo by design — the binaural pair
 *  is the whole point — so anything carrying it must stay in two. */
export function writeWavStereo(left: Float32Array, right: Float32Array,
                               path: string, rate = sampleRate): void {
  const n = Math.min(left.length, right.length);
  const body = Buffer.alloc(n * 4);
  for (let i = 0; i < n; i++) {
    body.writeInt16LE(toPCM(left[i]!), i * 4);
    body.writeInt16LE(toPCM(right[i]!), i * 4 + 2);
  }
  writeFileSync(path, Buffer.concat([header(2, n, rate), body]));
}

/**
 * Measuring rendered audio, so a broken render can be *found* rather than
 * listened for. Ported from `AudioProbe` in `AudioIO.swift`.
 */
export interface RenderQuality {
  seconds: number;
  peak: number;
  clippedSamples: number;
  nonFiniteSamples: number;
  leadingQuietSeconds: number;
  trailingQuietSeconds: number;
}

/** **A small, mechanical contract.** It does not claim the speech sounds good;
 *  it rejects the file-level defects that can be proven without pretending an
 *  acoustic metric has ears. */
export function isSafe(q: RenderQuality, edgeQuiet: number): boolean {
  return q.seconds > 0 && q.clippedSamples === 0 && q.nonFiniteSamples === 0
    && q.leadingQuietSeconds >= edgeQuiet * 0.95
    && q.trailingQuietSeconds >= edgeQuiet * 0.95;
}

/** Peak, clipping and the two file edges of a rendered speech unit.
 *  `preparedSpeechPart` guarantees the edge quiet; this reads it back, so a
 *  later refactor cannot silently remove the guarantee. */
export function renderQuality(samples: Float32Array, rate = sampleRate,
                              quietThreshold = 0.005): RenderQuality {
  let peak = 0, clipped = 0, nonFinite = 0;
  for (const sample of samples) {
    if (!Number.isFinite(sample)) { nonFinite++; continue; }
    const magnitude = Math.abs(sample);
    if (magnitude > peak) peak = magnitude;
    if (magnitude >= 0.999) clipped++;
  }
  let leading = 0;
  while (leading < samples.length && Number.isFinite(samples[leading]!)
         && Math.abs(samples[leading]!) < quietThreshold) leading++;
  let trailing = 0;
  while (trailing < samples.length
         && Number.isFinite(samples[samples.length - 1 - trailing]!)
         && Math.abs(samples[samples.length - 1 - trailing]!) < quietThreshold) trailing++;
  return {
    seconds: samples.length / rate,
    peak, clippedSamples: clipped, nonFiniteSamples: nonFinite,
    leadingQuietSeconds: leading / rate,
    trailingQuietSeconds: trailing / rate,
  };
}
